### PR TITLE
Dimension table storage quota config and validation

### PR DESCRIPTION
**Pull request by cbalci**
## Description
Adding some validation logic to ensure Dimension table sizes are properly constrained. This is a follow up to the [introduction of Dimension tables](https://github.com/apache/incubator-pinot/pull/6286) and https://github.com/apache/incubator-pinot/pull/6346 . In current implementation, dimension tables are fully loaded into heap memory, so this change is to prevent users from accidentally loading very large tables and disrupting the server.
Changes include:
- New controller config key: "controller.dimTable.maxSize"
- Default value if no config is provided: DEFAULT_DIM_TABLE_MAX_SIZE = "200M"
- Validation and enforcement logic in table creation flow (`PinotTableRestletResource`), such that the default max value is applied if no storage config exists for dimension tables. 

**Note:** I looked into enforcing this constraint in `TableConfigUtils.validate()` however it didn't work out due to package dependency issues. We need `ControllerConf` values for the validation but `TableConfigUtils` cannot depend on `ControllerConf` due to package structure.

## Testing
- Controller tests are updated to include different scenarios.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 1]
```diff
@@ -145,6 +145,7 @@ private static long getRandomInitialDelayInSeconds() {
   // If it's set to false, existing HLC realtime tables will stop consumption, and creation of new HLC tables will be disallowed.
   // Please make sure there is no HLC table running in the cluster before disallowing it.
   public static final String ALLOW_HLC_TABLES = "controller.allow.hlc.tables";
+  private static final String DIM_TABLE_MAX_SIZE = "controller.dimTable.maxSize";
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:148]: Make the config key public

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 2]
```diff
@@ -463,6 +467,36 @@ private void ensureMinReplicas(TableConfig tableConfig) {
     }
   }
 
+  private void ensureStorageQuotaConstraints(TableConfig tableConfig) {
+    // Dim tables must adhere to cluster level storage size limits
+    if (tableConfig.isDimTable()) {
+      QuotaConfig quotaConfig = tableConfig.getQuotaConfig();
+      String maxAllowedSize = this._controllerConf.getDimTableMaxSize();
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:474]: (Convention) Remove `this.`

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 3]
```diff
@@ -463,6 +467,36 @@ private void ensureMinReplicas(TableConfig tableConfig) {
...
+      long maxAllowedSizeInBytes = DataSizeUtils.toBytes(maxAllowedSize);
+
+      if (quotaConfig == null) {
+        // set a default storage quota
+        LOGGER.info("Assigning default storage quota for dimension table: {}", tableConfig.getTableName());
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:479]: Put the default value in the info message

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 4]
```diff
@@ -463,6 +467,36 @@ private void ensureMinReplicas(TableConfig tableConfig) {
...
+        tableConfig.setQuotaConfig(
+            new QuotaConfig(maxAllowedSize, null));
+      } else {
+        if (quotaConfig.getStorage() == null) {
+          // set a default storage quota and keep the RPS value
+          LOGGER.info("Assigning default storage quota for dimension table: {}", tableConfig.getTableName());
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:485]: Same here

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 5]
```diff
@@ -463,6 +467,36 @@ private void ensureMinReplicas(TableConfig tableConfig) {
...
+          tableConfig.setQuotaConfig(
+              new QuotaConfig(maxAllowedSize, quotaConfig.getMaxQueriesPerSecond())
+          );
+        } else {
+          if (quotaConfig.getStorageInBytes() > maxAllowedSizeInBytes) {
+            throw new PinotHelixResourceManager.InvalidTableConfigException(
+                "Invalid value for table storage quota. Max allowed is " + maxAllowedSize
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:492]: (nit)
```suggestion
                String.format("Invalid storage quota: %d, max allowed size: %d", storageInBytes, maxAllowedSize)
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

**Comment by cbalci** [comment]
Thanks for the review @Jackie-Jiang !

**Pull request merged**